source/microbit/modradio.cpp

```diff
@@ -0,0 +1,275 @@
+
+        // if the CRC was valid then accept the packet
+        if (NRF_RADIO->CRCSTATUS == 1) {
+            //printf("rssi: %d\r\n", -NRF_RADIO->RSSISAMPLE);
+
+            // only move the rx_buf pointer if there is enough room for another full packet
+            if (rx_buf + 1 + len + 1 + max_len <= buf_end) {
+                rx_buf += 1 + len;
+                NRF_RADIO->PACKETPTR = (uint32_t)rx_buf;
+            }
+        }
+
+        NRF_RADIO->TASKS_START = 1;
+    }
+}
+
+static void ensure_enabled(void) {
+    if (buf_start == NULL) {
+        nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "radio is not enabled"));
+    }
+}
+
+static void radio_disable(void) {
+    NVIC_DisableIRQ(RADIO_IRQn);
+    NRF_RADIO->EVENTS_DISABLED = 0;
+    NRF_RADIO->TASKS_DISABLE = 1;
+    while (NRF_RADIO->EVENTS_DISABLED == 0);
+    // free any old buffers
+    if (buf_start != NULL) {
+        m_del(uint8_t, buf_start, buf_end - buf_start);
+        buf_start = NULL;
+    }
+}
+
+static void radio_enable(size_t max_payload, size_t queue_len) {
+    radio_disable();
+
+    // allocate tx and rx buffers
+    queue_len += 1; // one extra for tx buffer
+    buf_start = m_new(uint8_t, (max_payload + 1) * queue_len);
+    buf_end = buf_start + (max_payload + 1) * queue_len;
+    rx_buf = buf_start + max_payload + 1; // start is tx buffer
+
+    // Enable the High Frequency clock on the processor. This is a pre-requisite for
+    // the RADIO module. Without this clock, no communication is possible.
+    NRF_CLOCK->EVENTS_HFCLKSTARTED = 0;
+    NRF_CLOCK->TASKS_HFCLKSTART = 1;
+    while (NRF_CLOCK->EVENTS_HFCLKSTARTED == 0);
+
+    // power should be one off: -30, -20, -16, -12, -8, -4, 0, 4
+    NRF_RADIO->TXPOWER = 0;
+
+    // should be between 0 and 100 inclusive (actual physical freq is 2400MHz + this register)
+    NRF_RADIO->FREQUENCY = 7;
+
+    // Configure for 1Mbps throughput.
+    // This may sound excessive, but running a high data rates reduces the chances of collisions...
+    NRF_RADIO->MODE = RADIO_MODE_MODE_Nrf_1Mbit;
+
+    // Configure the addresses we use for this protocol. We run ANONYMOUSLY at the core.
+    // A 40 bit addresses is used. The first 32 bits match the ASCII character code for "uBit".
+    // We also map the assigned 8-bit GROUP id into the PREFIX field. This allows the RADIO hardware to perform
+    // address matching for us, and only generate an interrupt when a packet matching our group is received.
+    NRF_RADIO->BASE0 = 0x75626974;
+
+    // Join the default group. This will configure the remaining byte in the RADIO hardware module.
+    NRF_RADIO->PREFIX0 = 0;
+
+    // The RADIO hardware module supports the use of multiple addresses, but as we're running anonymously, we only need one.
+    // Configure the RADIO module to use the default address (address 0) for both send and receive operations.
+    NRF_RADIO->TXADDRESS = 0;
+    NRF_RADIO->RXADDRESSES = 1;
```

Comment:
It is not very clear to me what is the difference between `TXADDRESS ` and `RXADDRESSES ` apart from the TX/RX. Does "addresses" indicate the number of addresses that the radio should pick up?